realm/src/main/java/io/realm/BaseRealm.java

```diff
@@ -586,14 +626,278 @@ public static synchronized void migrateRealm(RealmConfiguration configuration, R
     private class RealmCallback implements Handler.Callback {
         @Override
         public boolean handleMessage(Message message) {
-            if (message.what == REALM_CHANGED) {
-                sharedGroupManager.advanceRead();
-                sendNotifications();
+            switch (message.what) {
```

Comment:
In general this is really complicated to get a solid overview of, and could IMO use a refactoring into more digestible methods. I have a nasty feeling it to be quite a lot of methods, so would it make sense to introduce something like a `AsyncController` class that could contain all this logic? It would also make easier to split it into smaller methods without adding to many small private methods to the BaseRealm class.

IMO, the most important thing when scanning the code is being able to get a grasp of the control structure, and this has a lot of nested levels.

What do you think?